progress/__init__.py

```diff
@@ -65,19 +65,20 @@ def start(self):
     def finish(self):
         pass
 
-    def next(self, n=1):
-        if n > 0:
-            now = time()
-            dt = now - self._ts
-            if dt < self.time_threshold:
-                self._pending += n
-            else:
-                self._xput.append((n + self._pending) / dt)
-                self._ts = now
-                self._pending = 0
 
+    def next(self, n=1):
+        if n <= 0: return
+        now = time()
+        dt = now - self._ts
+        self._pending += n
+        #avoid performing computationally intensive tasks more than 1/dt (10 by default) times a second
+        if dt > self.time_threshold:
+            self._xput.append((self._pending) / dt)
+            self._ts = now
+            self._pending = 0
+            self.update()
         self.index = self.index + n
-        self.update()
+
 
```

Comment:
So you did 3 changes here:
1) Bail out early if n <= 0 (good)
2) Flattened the time_threshold check (I don't see the reason for this, they are 2 distinct cases, and you are changing self._pending just to set it to 0 later)
3) Call update() only when we are passed the threshold (good)

I would keep 1 and 3, but I don't see any gains with 2.